woocommerce/woocommerce
# Validate shipping rates

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





- This is essentially a replay of https://github.com/woocommerce/woocommerce/pull/37847.
- The above PR has *already been reviewed and tested,* but a test fail (leaky state from an older, existing test) was blocking merge.
- That fail has hopefully been addressed by https://github.com/woocommerce/woocommerce/pull/39136, however I was unable to rebase https://github.com/woocommerce/woocommerce/pull/37847 directly ... hence, I have created this replacement PR (which still contains the same commits by the same author) to move things along.

Closes https://github.com/woocommerce/woocommerce/issues/27030.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

- Navigate to **WooCommerce ▸ Settings ▸ Shipping Classes** and define two classes. For example:
    - Delicate
    - Perishable
- Now go to **WooCommerce ▸ Settings ▸ Shipping Zones** and setup a zone (example: Canada) with two flat rate shipping methods. Each method should support a single class. For example, in relation to the Delicate shipping class, we could have:
    - Flat rate
    - Delicate shipping class cost **$5**
    - Calculation type **per class**
    - Leave all other fields empty/set to their default value

<div align="center"><img width="780" alt="Screenshot 2023-07-07 at 06 21 43" src="https://github.com/woocommerce/woocommerce/assets/3594411/95df4c00-bf8c-46fd-869c-6997ac59475e"></div>

- Now create two products (or edit two existing ones).
    - Apply "Delicate" shipping class to one.
    - Apply "Perishable" to the other.
- Acting as a customer, add both items to the cart and try to checkout (and be sure to specify an appropriate address—such as in Canada, assuming you followed the above steps, so that you match our test shipping zones).
    - With this change, it should not be possible to checkout (because we don't have any shipping methods that support both classes). This is shown in the screenshot below.
    - Remove an item from the cart, and it should become possible to checkout (because we should have a shipping method matching the remaining product class).

<div align="center"><img width="340" alt="Screenshot 2023-07-07 at 06 30 14" src="https://github.com/woocommerce/woocommerce/assets/3594411/bcaf2ace-e1ee-447a-970d-f3750b8a3959"></div>

This is different from the behavior in the existing version of WooCommerce, in which it was possible to checkout with both items in the cart. 




### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
